clarity/contracts/arkadiko-liquidation-rewards-v1-1.clar

```diff
@@ -0,0 +1,186 @@
+;; ---------------------------------------------------------
+
+;; @desc add rewards to the contract
+;; @param share-block; block on which user shares in USDA pool are checked
+;; @param token; the reward token
+;; @param total-amount; amount of rewards
+(define-public (add-reward (share-block uint) (token-is-stx bool) (token <ft-trait>) (total-amount uint))
+  (let (
+    (reward-id (var-get total-reward-ids))
+  )
+    (asserts! (not (get-shutdown-activated)) (err ERR-EMERGENCY-SHUTDOWN-ACTIVATED))
+
+    ;; Transfer rewards to contract
+    (if token-is-stx
+      (try! (stx-transfer? total-amount tx-sender (as-contract tx-sender)))
+      (try! (contract-call? token transfer total-amount tx-sender (as-contract tx-sender) none))
+    )
+
+    ;; Add reward to map
+    (map-set reward-data  { reward-id: reward-id } { 
+      share-block: share-block,
+      token-is-stx: token-is-stx,
+      token: (contract-of token),
+      total-amount: total-amount
+    })
+
+    ;; Update total-reward-ids
+    (var-set total-reward-ids (+ reward-id u1))
+
+    (ok true)
+  )
+)
+
+;; ---------------------------------------------------------
+;; Get rewards
+;; ---------------------------------------------------------
+
+;; @desc get amount of rewards for user
+;; @param user; user to get rewards for
+;; @param reward-id; the rewards
+(define-public (get-rewards-of (user principal) (reward-id uint) (liquidation-pool <liquidation-pool-trait>))
+  (let (
+    (reward-info (get-reward-data reward-id))
+    (share-block (get share-block reward-info))
+    (total-amount (get total-amount reward-info))
+
+    (user-shares (unwrap-panic (contract-call? liquidation-pool get-shares-at user share-block)))
+    (total-rewards (/ (* user-shares total-amount) u10000000))
+
+    (claimed (get claimed-amount (get-reward-claimed reward-id user)))
+  )
+    (asserts! (is-eq (contract-of liquidation-pool) (unwrap-panic (contract-call? .arkadiko-dao get-qualified-name-by-name "liquidation-pool"))) (ok u0))
+
+    (ok (- total-rewards claimed))
+  )
+)
+
+;; @desc claim rewards for user
+;; @param reward-id; the rewards
+;; @param token; the reward token
+;; @param liquidation-pool; the pool on which shares are based
+(define-public (claim-rewards-of (reward-id uint) (token <ft-trait>) (liquidation-pool <liquidation-pool-trait>))
+  (let (
+    (user tx-sender)
+    (reward-amount (unwrap-panic (get-rewards-of user reward-id liquidation-pool)))
+    (reward-info (get-reward-data reward-id))
+  )
+    (asserts! (not (get-shutdown-activated)) (err ERR-EMERGENCY-SHUTDOWN-ACTIVATED))
+    (asserts! (is-eq (contract-of token) (get token reward-info)) (err ERR-WRONG-TOKEN))
+
+    (if (is-eq reward-amount u0)
+      (ok u0)
+      (begin 
+        (if (get token-is-stx reward-info)
+          (try! (as-contract (stx-transfer? reward-amount (as-contract tx-sender) user)))
+          (try! (as-contract (contract-call? token transfer reward-amount (as-contract tx-sender) user none)))
```

Comment:
REQ: I think this is a severe vulnerability, as follows

1. Anyone can add any SIP10 token, meaning that I could implement a malicious FT that transfers a bunch of STX in its transfer function.
2. I start staking in the pool and add 1 token of this as a reward (`add-reward` function)
3. I wait until the epoch is over
4. I run this function (`claim-rewards-of`)
5. Since I am eligible for the rewards, my malicious transfer function will run and drain all the tokens that are kept on this contract (could be xBTC, STX, ...)

Reply:
Solution is simple

Add a map with key principal `(contract-of token)` that keeps a whitelist and checks that in an assertion when adding and claiming rewards

Reply:
Right, great find! Will apply the solution.